### PR TITLE
Additional ORC writer fixes

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

**Comment by martint** [comment]
First and third commit look good.

**Comment by haozhun** [comment]
* [ ] [db420dcba0](https://github.com/prestodb/presto/pull/8861/commits/db420dcba0) Add OrcEncoding enum to differentiate between ORC and DWRF
* [ ] [d74a7b08ce](https://github.com/prestodb/presto/pull/8861/commits/d74a7b08ce) Cleanup OrcMetadataReader warnings
* [ ] [7d7a7276bf](https://github.com/prestodb/presto/pull/8861/commits/7d7a7276bf) Update to latest ORC proto IDL
* [ ] [bd8ecc1dd1](https://github.com/prestodb/presto/pull/8861/commits/bd8ecc1dd1) Recalculate stats during write validation
* [ ] [ad75faa1b9](https://github.com/prestodb/presto/pull/8861/commits/ad75faa1b9) Add configuration property to enable ORC writer validation
* [ ] [212e03c725](https://github.com/prestodb/presto/pull/8861/commits/212e03c725) Add ORC IntegerStatistics sum
* [ ] [71852b34e0](https://github.com/prestodb/presto/pull/8861/commits/71852b34e0) Ensure ORC StringStatistics sum does not overflow
* [ ] [892d398da3](https://github.com/prestodb/presto/pull/8861/commits/892d398da3) Add ORC writer JMX stats
* [ ] [bd3dbff067](https://github.com/prestodb/presto/pull/8861/commits/bd3dbff067) Move OrcWriter config to OrcWriterOptions
* [ ] [4d0996a6de](https://github.com/prestodb/presto/pull/8861/commits/4d0996a6de) Add Hive configuration for ORC writer options
* [ ] [eaf5b43a8b](https://github.com/prestodb/presto/pull/8861/commits/eaf5b43a8b) Align ORC default stripe size with open source Hive
* [ ] [df39e2a4e5](https://github.com/prestodb/presto/pull/8861/commits/df39e2a4e5) Inclide dictionary bytes in stripe flush decission

**Comment by haozhun** [comment]
* [db420dcba0](https://github.com/prestodb/presto/pull/8861/commits/db420dcba0) Add OrcEncoding enum to differentiate between ORC and DWRF
  * looks good
* [d74a7b08ce](https://github.com/prestodb/presto/pull/8861/commits/d74a7b08ce) Cleanup OrcMetadataReader warnings
  * looks good
* [7d7a7276bf](https://github.com/prestodb/presto/pull/8861/commits/7d7a7276bf) Update to latest ORC proto IDL
  * looks good. It has typo in commit body: metadat -> metadata.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/IntegerStatisticsBuilder.java [hunk 1]
```diff
@@ -33,6 +36,13 @@ public void addValue(long value)
 
         minimum = Math.min(value, minimum);
         maximum = Math.max(value, maximum);
+
+        try {
```
Comment (haozhun) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/IntegerStatisticsBuilder.java]: I would only execute this try if `overflow` is not already set. Given the way this code is currently written, I'm concerned about repeated throws in a tight loop.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/IntegerStatisticsBuilder.java [hunk 2]
```diff
@@ -44,14 +54,23 @@ private void addIntegerStatistics(long valueCount, IntegerStatistics value)
         nonNullValueCount += valueCount;
         minimum = Math.min(value.getMin(), minimum);
         maximum = Math.max(value.getMax(), maximum);
+
+        if (value.getSum() != null) {
```
Comment (haozhun) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/IntegerStatisticsBuilder.java]: I would only execute this try if `overflow` is not already set. Given the way this code is currently written, I'm concerned about repeated throws in a tight loop.

File: presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestIntegerStatisticsBuilder.java [hunk 3]
```diff
@@ -61,4 +67,85 @@ public void testMinAverageValueBytes()
...
+        statisticsBuilder.addValue(-44);
+        statisticsList.add(statisticsBuilder.buildColumnStatistics());
+        assertMergedStringStatistics(statisticsList, 3, -44);
+
+        statisticsBuilder.addValue(100);
+        statisticsList.add(statisticsBuilder.buildColumnStatistics());
+        assertMergedStringStatistics(statisticsList, 6, (-44 * 2) + 100);
+    }
+
+    private static void assertMergedStringStatistics(List<ColumnStatistics> statisticsList, int expectedNumberOfValues, long expectedSum)
```
Comment (haozhun) [inline presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestIntegerStatisticsBuilder.java]: incorrect function name from copy-paste

File: presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestIntegerStatisticsBuilder.java [hunk 4]
```diff
@@ -61,4 +67,85 @@ public void testMinAverageValueBytes()
...
+        statisticsList.add(statisticsBuilder.buildColumnStatistics());
+        assertMergedStringStatistics(statisticsList, 1, 0);
+
+        statisticsBuilder.addValue(-44);
+        statisticsList.add(statisticsBuilder.buildColumnStatistics());
+        assertMergedStringStatistics(statisticsList, 3, -44);
+
+        statisticsBuilder.addValue(100);
+        statisticsList.add(statisticsBuilder.buildColumnStatistics());
+        assertMergedStringStatistics(statisticsList, 6, (-44 * 2) + 100);
```
Comment (haozhun) [inline presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestIntegerStatisticsBuilder.java]: Add test for merging null sum (and mixed sum?)

File: presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterFactory.java [hunk 5]
```diff
@@ -64,31 +67,39 @@
...
         this.typeManager = requireNonNull(typeManager, "typeManager is null");
         this.nodeVersion = requireNonNull(nodeVersion, "nodeVersion is null");
         this.hiveStorageTimeZone = requireNonNull(hiveStorageTimeZone, "hiveStorageTimeZone is null");
-        this.stats = requireNonNull(stats, "stats is null");
+        this.readStats = requireNonNull(readStats, "stats is null");
+    }
+
+    @Managed
+    @Flatten
+    public DateTimeZone getHiveStorageTimeZone()
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterFactory.java]: * (Before I realize this is likely a typo) Out of curiosity, where is JMX serialization for `DateTimeZone` defined?
* (Now that I realized this is likely a typo) Is this supposed to be `getStats` instead? I would run Presto manually with this and verify that the numbers looks reasonable with visualvm.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/IntegerStatisticsBuilder.java (hunk 2 above)
Comment (dain) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/IntegerStatisticsBuilder.java]: There is another problem here.  If the input "value" does not have a sum, then we can not output a sum.

**Comment by dain** [comment]
@haozhun updated.  back to you

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/IntegerStatisticsBuilder.java [hunk 6]
```diff
@@ -44,14 +54,29 @@ private void addIntegerStatistics(long valueCount, IntegerStatistics value)
...
         minimum = Math.min(value.getMin(), minimum);
         maximum = Math.max(value.getMax(), maximum);
+
+        if (value.getSum() == null) {
+            // if input value does not have a sum tag this stat as overflowed
+            // to prevent creation of the sum stats (since it was not provided
+            // for these values).
+            overflow = true;
+        }
+        else if (!overflow) {
```
Comment (haozhun) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/IntegerStatisticsBuilder.java]: The overflow check is also valuable to "add" (beside "merge" here)

**Pull request closed without merging**